### PR TITLE
EZP-30469: Fixed parallel Content publishing

### PR DESCRIPTION
**Pull request by pawbuj**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30469](https://jira.ez.no/browse/EZP-30469)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | master
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

This PR resolves an edge case where two Editors publish different Versions of the same Content item at the same time.

This PR also creates base for parallel integration testing of Repository services, for now for internal use by the Repository.

## Background

The process of publishing archives recently published content item, therefore at the moment of update there should be no published version. This is the basis for both blocking publishing of another version and assessment if someone else published Content at the same time.

## QA

Testing this manually is quite tricky and requires at least two people working on a shared instance of eZ Platform.
1. Create Content item.
2. Create one draft from the first Version.
3. Create another draft from the first Version.
4. Publish both drafts simultaneously.
5. Observe that there are two Versions with "published" status.

Expected behavior: Observe for one Editor an error stating that someone else just published that Content item. Editor should be able to publish his Version then, keeping in mind that this will override someone else's work.

Alternatively this can be tested by one person by setting breakpoint after the first `setStatus` call in `\eZ\Publish\Core\Persistence\Legacy\Content\Handler::publish`.

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -55,7 +55,8 @@
         "friendsofphp/php-cs-fixer": "~2.15.1",
         "phpunit/phpunit": "^8.2",
         "matthiasnoback/symfony-dependency-injection-test": "^4.0",
-        "ezsystems/behatbundle": "^7.0@dev"
+        "ezsystems/behatbundle": "^7.0@dev",
+        "jenner/simple_fork": "^1.2"
```
Comment (alongosz) [inline composer.json]: Not important, nitpick: I'd move it either after `brianium/paratest` or at least one line above, so the diff does not affect unrelated behat bundle dependency.

File: eZ/Publish/API/Repository/Tests/Parallel/ParallelProcessList.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+{
+    /** @var Jenner\SimpleFork\Process[] */
+    private $pool = [];
+
+    public function addProccess(Process $callback): void
+    {
+        $this->pool[] = $callback;
+    }
+
+    public function getIterator()
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/Parallel/ParallelProcessList.php]:
```suggestion
    public function getIterator(): \Iterator
```
or something more relevant e.g `Traversable`?

File: eZ/Publish/API/Repository/Tests/Parallel/BaseParallelTestCase.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        $connection = $this->getRawDatabaseConnection();
+
+        $proccess = new Process(function () use ($callback, $connection) {
+            $connection->connect();
+            $callback();
+            $connection->close();
+        });
+
+        $list->addProccess($proccess);
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/Parallel/BaseParallelTestCase.php]:
```suggestion
        $list->addProcess($process);
```

File: eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php [hunk 4]
```diff
@@ -602,6 +602,65 @@ public function setStatus($contentId, $version, $status)
...
 
+    public function setPublishedStatus(int $contentId, int $versionNo): void
+    {
+        $query = $this->getSetVersionStatusQuery(
+            $contentId,
+            $versionNo,
+            VersionInfo::STATUS_PUBLISHED
+        );
+
+        $notExistPublishedVersion = <<< HEREDOC
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php]: Maybe add some inline doc here to explain what this does to go easy on us all in the future, as it contains multiple levels of SQL logic it's not easily readable what it intends to do 😉

**Comment by alongosz** [comment]
> @alongosz There is still a _(heavily reduced)_ risk of race condition between when update to `ezcontentobject_version` table and `ezcontentobject` table is done right?

Well... Yes, but regardless of which process updates `ezcontentobject` table, the result is the same.

> Anyway you see we can improve that further in the future?
> 
> _(E.g. by optimizing publish further and doing more in one SPI call where we do some locking or something to make sure there is absolutely no risk of race conditions for the whole current version swap)_

Indeed, @pawbuj also suggested to use locks (e.g. Symfony Lock). This might be something to consider. For me consistency of atomic update felt more important. This is in line with doing more in one SPI call, so, yes we could explore this as well in the future.

File: eZ/Publish/Core/Persistence/Legacy/Content/Gateway/ExceptionConversion.php [hunk 5]
```diff
@@ -172,6 +173,17 @@ public function setStatus($contentId, $version, $status)
         }
     }
 
+    public function setPublishedStatus(int $contentId, int $status): void
+    {
+        try {
+            $this->innerGateway->setPublishedStatus($contentId, $status);
+        } catch (BadStateException $e) {
```
Comment (adamwojs) [inline eZ/Publish/Core/Persistence/Legacy/Content/Gateway/ExceptionConversion.php]: This statement is redundant:

```php
} catch (BadStateException $e) {
    throw $e;
}
```

File: eZ/Publish/API/Repository/Tests/Parallel/ParallelProcessList.php [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+namespace eZ\Publish\API\Repository\Tests\Parallel;
+
+use Jenner\SimpleFork\Process;
+
+class ParallelProcessList implements \IteratorAggregate
+{
+    /** @var \Jenner\SimpleFork\Process[] */
+    private $pool = [];
+
+    public function addProcess(Process $callback): void
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Tests/Parallel/ParallelProcessList.php]:
```suggestion
    public function addProcess(Process $process): void
```

File: eZ/Publish/API/Repository/Tests/Parallel/ParallelProcessList.php [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\API\Repository\Tests\Parallel;
+
+use Jenner\SimpleFork\Process;
+
+class ParallelProcessList implements \IteratorAggregate
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Tests/Parallel/ParallelProcessList.php]:
```suggestion
final class ParallelProcessList implements \IteratorAggregate
```

File: eZ/Publish/API/Repository/Tests/Parallel/BaseParallelTestCase.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+abstract class BaseParallelTestCase extends BaseTest
+{
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $connection = $this->getRawDatabaseConnection();
+        $dbms = $connection->getDatabasePlatform()->getName();
+
+        if (!in_array($dbms, ['mysql', 'postgresql'])) {
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Tests/Parallel/BaseParallelTestCase.php]:
```suggestion
        if ($dbms !== `sqlite`) {
```

File: eZ/Publish/API/Repository/Tests/Parallel/BaseParallelTestCase.php [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            $connection->close();
+        });
+
+        $list->addProcess($process);
+    }
+
+    protected function runParallelProcesses(ParallelProcessList $list): void
+    {
+        $connection = $this->getRawDatabaseConnection();
+        $connection->close();
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Tests/Parallel/BaseParallelTestCase.php]: IMHO this requires comment with some explanation 😉

File: eZ/Publish/API/Repository/Tests/Parallel/ContentServiceTest.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\API\Repository\Tests\Parallel;
+
+use eZ\Publish\Core\Base\Exceptions\BadStateException;
+
+class ContentServiceTest extends BaseParallelTestCase
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Tests/Parallel/ContentServiceTest.php]:
```suggestion
final class ContentServiceTest extends BaseParallelTestCase
```

File: eZ/Publish/API/Repository/Tests/Parallel/BaseParallelTestCase.php (hunk 9 above)
Comment (pawbuj) [reply eZ/Publish/API/Repository/Tests/Parallel/BaseParallelTestCase.php]: https://www.php.net/manual/en/function.pcntl-fork.php#70721

Comment (alongosz) [reply eZ/Publish/API/Repository/Tests/Parallel/BaseParallelTestCase.php]: > IMHO this requires comment with some explanation
>
> https://www.php.net/manual/en/function.pcntl-fork.php#70721

Yep, let's add this link as an inline `// @see` comment.

**Comment by andrerom** [comment]
> Indeed, @pawbuj also suggested to use locks (e.g. Symfony Lock). This might be something to consider. For me consistency of atomic update felt more important. This is in line with doing more in one SPI call, so, yes we could explore this as well in the future.

👍 Great that it's on the radar!

Adding Symfony lock to this part could be a way, but it adds *a lot* of complexity which might be better solved by having row lock in SQL rather or otherwise _(to avoid having to orchestrate this in PHP doing ping pong to SQL, Cache and now also distributed Lock system)_. But something like that would require moving more of the logic to StorageEngine for the 3 version+content status swap to be fully atomic*.

_* With that there would be no need to throw on conflict, just let the last one win no matter what on this low level. High level publish locking can rather be a feature for workflow engine to make sure two versions can't reach the same final stage before publishing for instance._

**Comment by mnocon** [comment]
I was able to reproduce the issue and confirm the fix with xdebug (we've tried invoking it together with @katarzynazawada, but we had no luck).

Also performed some checks around publishing, QA approved.

**Comment by lserwatka** [comment]
You can merge it up.

**Comment by pawbuj** [comment]
done
